File: client/src/lib/fs.ts
import { getBridgeClientRequestor } from '@/lib/utils'

export type Stats = {
  size: number
  isDir: boolean
  isFile: boolean
  modifiedTime: number
}

export type Action =
  | 'stats'
  | 'read-dir'
  | 'create-dir'
  | 'rename'
  | 'remove'
  | 'copy-file'
  | 'read-text-file'
  | 'write-text-file'
  | 'download-file'
  | 'read-archive'

export class Client {
  private request = getBridgeClientRequestor<Action>('fs')

  async stats(path: string): Promise<Stats> {
    return await this.request({ action: 'stats', data: { path } })
  }

  async readDir(path: string): Promise<string[]> {
    return await this.request({ action: 'read-dir', data: { path } })
  }

  async createDir(path: string, recursive?: boolean): Promise<void> {
    return await this.request({ action: 'create-dir', data: { path, recursive } })
  }

  async rename(src: string, dst: string): Promise<void> {
    return await this.request({ action: 'rename', data: { src, dst } })
  }

  async remove(path: string, recursive?: boolean): Promise<void> {
    return await this.request({ action: 'remove', data: { path, recursive } })
  }

  async copyFile(src: string, dst: string): Promise<void> {
    return await this.request({ action: 'copy-file', data: { src, dst } })
  }

  async readTextFile(path: string): Promise<string> {
    return await this.request({ action: 'read-text-file', data: { path } })
  }

  async writeTextFile(path: string, data: string): Promise<void> {
    return await this.request({ action: 'write-text-file', data: { path, data } })
  }

  async downloadFile(url: string, path: string): Promise<void> {
    return await this.request({ action: 'download-file', data: { url, path } })
  }

  async readArchive(path: string, container: string): Promise<string[]> {
    return await this.request({ action: 'read-archive', data: { path, container } })
  }
}

export function getFileNameExtension(value: string) {
  return value.slice(((value.lastIndexOf('.') - 1) >>> 0) + 2)
}

File: backend/src/routes/bridges/fs.rs
use crate::error::{Error, Result};
use crate::fs::{download_file, read_archive, stats};
use crate::utils::{BridgeRequest, State};
use axum::Json;
use axum::extract::State as StateExtractor;
use axum::response::IntoResponse;
use serde_json::{Value as JsonValue, from_value as from_json, to_value as to_json};
use std::fs;
use std::path::PathBuf;

pub async fn handler(
    state: StateExtractor<State>,
    Json(mut req): Json<BridgeRequest<JsonValue>>,
) -> Result<impl IntoResponse> {
    match req.action.as_str() {
        "stats" => {
            let path: PathBuf = from_json(req.data["path"].take())?;
            let res = stats(&path)?;

            Ok(Json(to_json(&res)?))
        }

        "read-dir" => {
            let path: PathBuf = from_json(req.data["path"].take())?;
            let mut res = vec![];

            for entry in fs::read_dir(path)? {
                res.push(entry?.file_name().to_string_lossy().to_string());
            }

            Ok(Json(to_json(&res)?))
        }

        "create-dir" => {
            let path: PathBuf = from_json(req.data["path"].take())?;
            let recursive: bool = from_json(req.data["recursive"].take()).unwrap_or(false);

            if recursive {
                fs::create_dir_all(path)?
            } else {
                fs::create_dir(path)?
            }

            Ok(Json(JsonValue::Null))
        }

        "rename" => {
            let src: PathBuf = from_json(req.data["src"].take())?;
            let dst: PathBuf = from_json(req.data["dst"].take())?;

            fs::rename(src, dst)?;
            Ok(Json(JsonValue::Null))
        }

        "remove" => {
            let path: PathBuf = from_json(req.data["path"].take())?;
            let recursive: bool = from_json(req.data["recursive"].take()).unwrap_or(false);

            if recursive {
                fs::remove_dir_all(path)?
            } else {
                fs::remove_file(path)?
            }

            Ok(Json(JsonValue::Null))
        }

        "copy-file" => {
            let src: PathBuf = from_json(req.data["src"].take())?;
            let dst: PathBuf = from_json(req.data["dst"].take())?;

            fs::copy(src, dst)?;
            Ok(Json(JsonValue::Null))
        }

        "read-text-file" => {
            let path: PathBuf = from_json(req.data["path"].take())?;
            let res = fs::read_to_string(path)?;

            Ok(Json(to_json(&res)?))
        }

        "write-text-file" => {
            let path: PathBuf = from_json(req.data["path"].take())?;
            let data: String = from_json(req.data["data"].take())?;

            fs::write(path, data)?;
            Ok(Json(JsonValue::Null))
        }

        "download-file" => {
            let url: String = from_json(req.data["url"].take())?;
            let path: PathBuf = from_json(req.data["path"].take())?;

            download_file(&state.http_client, &url, &path).await?;
            Ok(Json(JsonValue::Null))
        }

        "read-archive" => {
            let path: PathBuf = from_json(req.data["path"].take())?;
            let container: String = from_json(req.data["container"].take())?;
            let res = read_archive(&path, &container)?;

            Ok(Json(to_json(&res)?))
        }

        _ => Err(Error::new(format!("invalid action: {}", req.action))),
    }
}
